netket/utils/struct/dataclass.py

```diff
@@ -0,0 +1,320 @@
+
+        _cache = field(
+            pytree_node=pytree_node,
+            serialize=False,
+            default=Uninitialized,
+            repr=False,
+            hash=False,
+            init=True,
+            compare=False,
+        )
+        _set_new_attribute(clz, cache_name, _cache)
+        _set_new_attribute(clz, compute_name, method)
+        clz.__annotations__[cache_name] = return_type
+        # create accessor method
+        body_lines = [
+            f"if {self_name}.{cache_name} is Uninitialized:",
+            f"\tBUILTINS.object.__setattr__({self_name},{cache_name!r},self.{compute_name}())",
+            f"",
+            f"return {self_name}.{cache_name}",
+        ]
+
+        fun = _create_fn(
+            name, [self_name], body_lines, return_type=return_type, globals=globals
+        )
+        fun.__doc__ = method.__doc__
+        prop_fun = property(fun)
+        setattr(clz, name, prop_fun)
+
+        _precompute_body_method.append(f"{self_name}.{name}")
+
+    if len(_precompute_body_method) == 0:
+        _precompute_body_method.append("pass")
+
+    fun = _create_fn(name, [self_name], _precompute_body_method, globals=globals)
+    fun.__doc__ = "Precompute the value of all cached properties"
+    setattr(clz, "_precompute_cached_properties", fun)
+
+
+def attach_preprocess_init(data_clz, init_doc=MISSING):
+    preprocess_method_name = "__pre_init__"
+    dataclass_init_name = "__init_dataclass__"
+
+    if not preprocess_method_name in data_clz.__dict__:
+
+        def _preprocess_args_default(self, *args, **kwargs):
+            if hasattr(super(data_clz, self), preprocess_method_name):
+                args, kwargs = getattr(super(data_clz, self), preprocess_method_name)(
+                    *args, **kwargs
+                )
+
+            return args, kwargs
+
+        _set_new_attribute(
+            data_clz, preprocess_method_name, _preprocess_args_default
+        )  # lambda *args, **kwargs: args, kwargs)
+
+    _set_new_attribute(data_clz, dataclass_init_name, data_clz.__init__)
+
+    self_name = "self"
+    body_lines = [
+        f"if not __skip_preprocess:",
+        f"\targs, kwargs = {self_name}.{preprocess_method_name}(*args, **kwargs)",
+        f"{self_name}.{dataclass_init_name}(*args, **kwargs)",
+    ]
+
+    fun = _create_fn(
+        "__init__",
+        [self_name, "*args", "__skip_preprocess=False", "**kwargs"],
+        body_lines,
+    )
+    if init_doc is MISSING:
+        fun.__doc__ = getattr(data_clz, preprocess_method_name).__doc__
+    else:
+        fun.__doc__ = init_doc
+    setattr(data_clz, "__init__", fun)
+
+
+def dataclass(clz=None, *, init_doc=MISSING):
+    """
+    Decorator creating a NetKet-flavour dataclass.
+    This behaves as a flax dataclass, that is a Frozen python dataclass, with a twist!
+    See their documentation for standard behaviour.
+
+    The new functionalities added by NetKet are:
```

Comment:
just wondering: is there a chance we can push this to Flax instead of having to supporting it ourselves?

Reply:
Nope.
I asked but they are not interested in those fancy things.
Arguably, in a more 'functional' world like where flax is, those things are not needed.

But we live in somehow an hibrid world where our API is object oriented and the internals are not so....